### PR TITLE
Move `SortKeyCursor` and `RowIndex` into modules, add `sort_key_cursor` test

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

re : https://github.com/apache/arrow-datafusion/issues/2427

 # Rationale for this change
`SortKeyCursor::compare` is the bottleneck for merge performance (details on https://github.com/apache/arrow-datafusion/issues/2427#issuecomment-1140425072)

As I prepare to improve the performance of this code (and make it more complicated),  I want it more isolated and independent so I can:
1. Understand the code better
2. Validate my approach
3. Ensure I don't introduce performance regressions

# What changes are included in this PR?
1. Move `SortKeyCursor` into its own module
2. Move `RowIndex` into its own module
3. add `sort_key_cursor` test

# Are there any user-facing changes?
No

# Does this PR break compatibility with Ballista?
no

### CODE REVIEW

File: datafusion/core/src/physical_plan/sorts/cursor.rs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+use arrow::array::{ArrayRef, DynComparator};
+use arrow::compute::SortOptions;
+use arrow::record_batch::RecordBatch;
+use hashbrown::HashMap;
+use parking_lot::RwLock;
+use std::borrow::BorrowMut;
+use std::cmp::Ordering;
+use std::sync::Arc;
+
+/// A `SortKeyCursor` is created from a `RecordBatch`, and a set of
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sorts/cursor.rs:13]: This code is moved, and I added `stream_id()` and `batch_idx()` accessors

File: datafusion/core/src/physical_plan/sorts/index.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
+/// A `RowIndex` identifies a specific row in a logical stream.
+///
+/// Each stream is identified by an `stream_idx` and is formed from a
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sorts/index.rs:3]: This code is moved into a new module, and I added some comments / ascii art to help my future self not have to read the code the next time

File: datafusion/core/src/physical_plan/sorts/mod.rs [hunk 3]
```diff
@@ -17,208 +17,16 @@
...
+use crate::physical_plan::SendableRecordBatchStream;
 use std::fmt::{Debug, Formatter};
-use std::sync::Arc;
 
+mod cursor;
+mod index;
 pub mod sort;
 pub mod sort_preserving_merge;
 
-/// A `SortKeyCursor` is created from a `RecordBatch`, and a set of
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sorts/mod.rs:36]: This code is just moved

File: datafusion/core/tests/sort_key_cursor.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
+//! Contains tests for SortKeyCursor
```
Comment (alamb) [inline datafusion/core/tests/sort_key_cursor.rs:1]: These tests are new -- I plan to expand them over time

**Comment by alamb** [comment]
FYI @yjshen @richox

File: datafusion/core/src/physical_plan/sorts/index.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+/// │    │ │    │ │    │
+/// │ C1 │ │... │ │ CN │◀─────── (batch_idx = 1)
+/// │    │ │    │ │    │
+/// └────┘ └────┘ └────┘
+///
+///          ...
+///
+/// ┌────┐ ┌────┐ ┌────┐           RecordBatch
+/// │    │ │    │ │    │
+/// │ C1 │ │... │ │ CN │◀────── (batch_idx = N-1)
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/sorts/index.rs:44]:
```suggestion
/// │ C1 │ │... │ │ CN │◀────── (batch_idx = M-1)
```
I think the dimensions are independent?

File: datafusion/core/src/physical_plan/sorts/index.rs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+/// A `RowIndex` identifies a specific row in a logical stream.
+///
+/// Each stream is identified by an `stream_idx` and is formed from a
+/// sequence of RecordBatches batches, each of which is identified by
+/// a unique `batch_idx` within that stream.
+///
+/// This is used by `SortPreservingMergeStream` to identify which
+/// the order of the tuples in the final sorted output stream.
+///
+/// ```text
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/sorts/index.rs:27]: This diagram seems to imply all streams must have the same number of batches, I'm not sure this is the case

File: datafusion/core/tests/sort_key_cursor.rs [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+use std::{cmp::Ordering, sync::Arc};
+
+use arrow::{array::Int64Array, compute::SortOptions, record_batch::RecordBatch};
+use datafusion::physical_plan::sorts::{RowIndex, SortKeyCursor};
+use datafusion_physical_expr::expressions::col;
+
+/// Compares [`RowIndex`]es with a vector of strings, the result of
+/// pretty formatting the [`RowIndex`]es. This is a macro so errors
```
Comment (tustvold) [inline datafusion/core/tests/sort_key_cursor.rs:27]: I mean unwrap() will give you a stack trace... I'm not a massive fan of macros as they are hard to read, and slow to compile...

File: datafusion/core/tests/sort_key_cursor.rs [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        "2: (0, 2)",
+        "2: (0, 3)",
+    ];
+
+    assert_indexes!(expected, run(&mut cursor1, &mut cursor2));
+}
+
+/// Runs the two cursors to completion, sorting them, and
+/// returning the sorted order of rows that would have produced
+fn run(cursor1: &mut SortKeyCursor, cursor2: &mut SortKeyCursor) -> Vec<RowIndex> {
```
Comment (tustvold) [inline datafusion/core/tests/sort_key_cursor.rs:86]: This code doesn't seem to be being used by SortPreservingMerge, is this intentional?

**Review by tustvold — APPROVED** [review]

File: datafusion/core/tests/sort_key_cursor.rs [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    let sort_key = vec![c1];
+
+    let sort_options = Arc::new(vec![SortOptions::default()]);
+
+    let mut cursor1 =
+        SortKeyCursor::new(1, 0, &batch1, &sort_key, Arc::clone(&sort_options)).unwrap();
+    let mut cursor2 =
+        SortKeyCursor::new(2, 0, &batch2, &sort_key, Arc::clone(&sort_options)).unwrap();
+
+    let expected = vec![
```
Comment (tustvold) [inline datafusion/core/tests/sort_key_cursor.rs:70]: If you were feeling fancy you could test stability, i.e. ties are broken by the lower stream idx

File: datafusion/core/tests/sort_key_cursor.rs (hunk 8 above)
Comment (alamb) [reply datafusion/core/tests/sort_key_cursor.rs:86]: At the moment, yes -- I want to unit test `SoryKeyCursor`. SortPreservingMerge already has reasonable test coverage in my mind

File: datafusion/core/src/physical_plan/sorts/index.rs (hunk 6 above)
Comment (alamb) [reply datafusion/core/src/physical_plan/sorts/index.rs:27]: that is an excellent point -- updated in 93668d2208

File: datafusion/core/src/physical_plan/sorts/index.rs (hunk 5 above)
Comment (alamb) [reply datafusion/core/src/physical_plan/sorts/index.rs]: in 93668d2208 👍

**Pull request merged**